### PR TITLE
canonicalize repository lcoation for storage path calculation

### PR DESCRIPTION
**Pull request by tomerd**
motivation: there is some built-in non-determinism where the graph resolution races and sometime one location variant wins in the location -> identity mapping, this can lead to looking up a different storage path since that does not use the normalized form of the location

changes:
* normalize (canonicalize) the repository location when computing the storage path
* add and adjust tests

rdar://112134161

### CODE REVIEW

**Comment by tomerd** [comment]
@swift-ci smoke test

**Comment by tomerd** [comment]
@swift-ci smoke test

File: Tests/WorkspaceTests/WorkspaceTests.swift [hunk 1]
```diff
@@ -3808,7 +3808,7 @@ final class WorkspaceTests: XCTestCase {
                     products: [
                         MockProduct(name: "Foo", targets: ["Foo"]),
                     ],
-                    versions: ["1.0.0"],
+                    versions: ["1.0.0", "1.1.0"],
```
Comment (MaxDesiatov) [inline Tests/WorkspaceTests/WorkspaceTests.swift:3811]: Is this change related?

**Review by MaxDesiatov — APPROVED** [review]

File: Tests/WorkspaceTests/WorkspaceTests.swift (hunk 1 above)
Comment (tomerd) [reply Tests/WorkspaceTests/WorkspaceTests.swift:3811]: yea it demonstrated the bug unintentionally

**Review by neonichu — APPROVED** [review]
